src/api/libcellml/logger.h

```diff
@@ -38,23 +39,75 @@ class LIBCELLML_EXPORT Logger
     Logger &operator=(Logger rhs) = delete; /**< Assignment operator */
 
     /**
-     * @brief Clear the errors from the logger.
+     * @brief Clear the issues from the logger.
      *
-     * Clear the errors from the logger.
+     * Clear the issues from the logger.
      */
-    void removeAllErrors();
+    void removeAllIssues();
 
     /**
-     * @brief Add an error to the logger.
+     * @brief Add an issue to the logger.
      *
-     * Adds the argument @p error to this logger.
+     * Adds the argument @p issue to this logger.
      *
-     * @param error The @c ErrorPtr to add.
+     * @param issue The @c IssuePtr to add.
      */
-    void addError(const ErrorPtr &error);
+    void addIssue(const IssuePtr &issue);
 
     /**
-     * @brief Get the number of errors.
+     * @brief Get the number of issues.
+     *
+     * Return the number of issues of any level currently stored in the logger.
+     *
+     * @return The number of issues.
+     */
+    size_t issueCount() const;
+
+    /**
+     * @brief Get issue of any level at the specified @p index.
+     *
+     * Returns an issue at the @p index.  If the @p index
+     * is not valid a @c nullptr is returned, the valid range for the @p index
+     * is [0, \#issues).
+     *
+     * @param index The index of the issue to return.
+     *
+     * @return A reference to the issue at the given index on success, @c nullptr otherwise.
+     */
+    IssuePtr issue(size_t index) const;
+
+    /**
+     * @brief Get issue of specified @p level at the specified @p index.
+     *
+     * Returns an issue at the @p index of specified @p level.  If the @p index
+     * is not valid a @c nullptr is returned, the valid range for the @p index
+     * is [0, \#issues_of_level).
+     *
+     * @param index The index of the issue to return.
+     * @param level The level of issue to return.
+     *
+     * @return A reference to the issue from the list of issues at the given level,
+     * at the given index on success, @c nullptr otherwise.
+     */
+    IssuePtr issue(size_t index, libcellml::Issue::Level level) const;
+
+    /**
+     * @brief Get issue of one of the specified @p levels at the specified @p index.
+     *
+     * Returns an issue at the @p index of one of the specified @p levels.  If the @p index
+     * is not valid a @c nullptr is returned, the valid range for the @p index
+     * is [0, \#issues_of_levels).
+     *
+     * @param index The index of the issue to return.
+     * @param levels The levels of issue to return.
+     *
+     * @return A reference to the issue from the list of issues at one of the given levels,
+     * at the given index on success, @c nullptr otherwise.
+     */
+    IssuePtr issue(size_t index, std::vector<libcellml::Issue::Level> &levels) const;
```

Comment:
I don't understand why you are passing in a vector of issue levels here?

Reply:
It was part of a request in the original discussion that people wanted to be able to retrieve, in order, a collection of more than one level of issue (eg, all `ERROR`s and `WARNING`s at once).

Reply:
Then I don't see why you have and index and not a flag used for the first parameter and I also don't see why you would return an `IssuePtr`. I also think the name of the method has to change because it is not retrieving an issue.

Reply:
The documentation doesn't describe what is going on at all.  It looks like a copy and paste from the method above.

Reply:
There are three functions which do similar things:
- `issue(myIndex)` will get the `issue` of any `level` at the `myIndex`.  It just searches all of them.  
- `issue(myIndex, myLevel) `will only search the one `level` specified by `myLevel`.
- `issue(myIndex, [ERROR, WARNING,...]) `will search all errors and warnings (or whatever is specified in the list) and return the issue at `myIndex`.  

Earlier discussions told me that it would be useful to be able to store the order in which issues were logged, and to be able to retrieve them in that same order even when more than one level of issue was wanted; the equivalent of setting the warnings-as-errors flag in a compiler.  I've updated the docstrings now so that they'll hopefully make this clearer?

Reply:
(see tests towards the end of the logger that show it works as expected)

Reply:
For me this method has to go.  It makes virtually no sense. If you want issues in order you get the issue count and iterate through the issues filtering what you do and don't want as you go.

Reply:
Can you maybe chat to @agarny about it?  This was something he specifically requested?

Reply:
To be able to retrieve a subset of issues (e.g. errors and warnings while skipping hints) is something that could indeed be useful. Now, I can't recall whether I explicitly asked for that exacty method. If I did then sorry.

Indeed, such a method is not optimal. If anything, I would want to go through a whole subset of issues. So, to call that method `n` times would be very inefficient.

Instead, I would expect a vector (containing the subset of issues in which I am interested) to be returned. From there I could efficiently go trhough them.

Reply:
I think it was that you wanted to be able to keep the order of errors and warnings the same, so that if you encountered error1, warning1, error2, warning2, hint1 they were reported e1, w1, e2 and not e1, e2, w1, w2.  If just getting one list with e1, e2, w1, w2 is ok then this method (and others!) can be substantially speeded up!!

Reply:
Yes, a list is what we need. Then, the consumer can do whatever s/he wants with it.

Reply:
@agarny Can I ask a favour? Can you have a look at the logger tests and let me know what you'd change there so that the issue fetching is how you intended it?  Once I have a clearer picture I can fix the problem ...

Reply:
Sure, but this will have to wait Wednesday (not around on Monday and Tuesday).

Reply:
I've removed it now in the interests of getting this one in.

Reply:
Ok, then I guess we will revisit that in a different issue/PR.